### ISSUE TITLE
mutation signatures change every time when rerun

### ISSUE THREAD

**Issue by lxilxi** [PoisonAlien/maftools#101]
I found the mutation signatures changed every time when I rerun it?  I thought for the same dataset, you would expect the mutation signature to the the same.  Here is my command:
lc.sign = extractSignatures(mat = lc.tnm, nTry = 6, plotBestFitRes = FALSE)

One run gave the following:

> Using 3 as a best-fit rank based on decreasing cophenetic correlation coefficient.
Comparing against experimentally validated 30 signatures.. (See http://cancer.sanger.ac.uk/cosmic/signatures for details.)
Found Signature_1 most similar to validated Signature_4. Aetiology: exposure to tobacco (smoking) mutagens [cosine-similarity: 0.708] 
Found Signature_2 most similar to validated Signature_6. Aetiology: defective DNA mismatch repair [cosine-similarity: 0.703] 
Found Signature_3 most similar to validated Signature_20. Aetiology: defective DNA mismatch repair [cosine-similarity: 0.535] 

Another run gave following:

> Using 3 as a best-fit rank based on decreasing cophenetic correlation coefficient.
Comparing against experimentally validated 30 signatures.. (See http://cancer.sanger.ac.uk/cosmic/signatures for details.)
Found Signature_1 most similar to validated Signature_20. Aetiology: defective DNA mismatch repair [cosine-similarity: 0.568] 
Found Signature_2 most similar to validated Signature_1. Aetiology: spontaneous deamination of 5-methylcytosine [cosine-similarity: 0.821] 
Found Signature_3 most similar to validated Signature_4. Aetiology: exposure to tobacco (smoking) mutagens [cosine-similarity: 0.785] 

Other runs gave various results.  Could you please help with this?  Thank you!

**Comment by PoisonAlien** [comment]
Hello ! 
Even I have observed this. It is possible that each run might produce slightly different results (due to seed value I guess), however overall results should be similar. I will try to see if I can fix it, but it depends on the underlying [nmf](https://cran.r-project.org/web/packages/NMF/index.html) on how it handles a fixed seed value.

**Comment by lxilxi** [comment]
Thank you for the information.  I thought seed was used to make reproducible results.  Probably nmf does it differently.  It would be great if it can be fixed, otherwise it's hard to explain to people.  Thanks again.

**Comment by PoisonAlien** [comment]
Yes, understandable. 
I think I know the issue. I only use seed while [estimating](https://github.com/PoisonAlien/maftools/blob/master/R/extractSignatures.R#L51) `n` but I never set it for [actual run](https://github.com/PoisonAlien/maftools/blob/master/R/extractSignatures.R#L75).  Will fix it soon. Thanks for reporting !

**Comment by lxilxi** [comment]
Awesome!  Thanks!

**Comment by PoisonAlien** [comment]
Hi,
I have pushed a commit to address this. Can you install the package from github and see if the runs are reproducible ? Please let me know if this works..

**Comment by lxilxi** [comment]
Thank you for the updated code.  Now the results are consistent with the same input file.  
One more question, there was a signature always appear with the code before this fix, now it's gone, would you give some insight?  Thank you!

**Comment by lxilxi** [comment]
I got another error when using extracSignatures function:
after >Estimating best rank..
It showed the similar metrics as before, but couldn't choose a number for rank:
> Using NA as a best-fit rank based on decreasing cophenetic correlation coefficient.
Error in if (rank < 1) fstop("invalid argument 'rank': must be greater than 0") : 
  missing value where TRUE/FALSE needed

**Comment by PoisonAlien** [comment]
Hi,
I am not sure about the missing signature, my guess would you are referring to signature6/20, which both are related defects in to DNA mismatch repair. Probably one of them is missing in updated results due to closer cosine similarity. 

Regarding your error, could you post your command ? I have never encountered it before. Do you still have the issue ?

**Comment by lxilxi** [comment]
Do you mean if one signature is close to several validated signatures, the assignment can change dramatically without setting the seed?   The missing signature might have similar characteristics to others,  but I thought each signature is unique with special traits.  Since now the results are reproducible, maybe I'll just use the new results.

**Comment by lxilxi** [comment]
I still have the Error, here is my command and error:
> lc2.sign = extractSignatures(mat = lc2.tnm, nTry = 6, plotBestFitRes = FALSE)
Warning : Found zero mutations for conversions G[C>G]AWarning : Found zero mutations for conversions G[T>A]AWarning : Found zero mutations for conversions G[T>A]CWarning : Found zero mutations for conversions T[T>A]CWarning : Found zero mutations for conversions G[T>C]GWarning : Found zero mutations for conversions T[T>C]AWarning : Found zero mutations for conversions T[T>C]GWarning : Found zero mutations for conversions A[T>G]AWarning : Found zero mutations for conversions C[T>G]AWarning : Found zero mutations for conversions C[T>G]CWarning : Found zero mutations for conversions G[T>G]CWarning : Found zero mutations for conversions G[T>G]TWarning : Found zero mutations for conversions T[T>G]A
Estimating best rank..
   method   seed rng metric rank sparseness.basis sparseness.coef      rss      evar
1: brunet random   5     KL    2        0.5470037       0.8881449 394.4189 0.2312721
2: brunet random   1     KL    3        0.5692186       0.9351232 360.6380 0.2971116
3: brunet random   4     KL    4        0.6046383       0.9308105 350.4350 0.3169973
4: brunet random   3     KL    5        0.6362677       0.9460098 320.5060 0.3753295
5: brunet random   2     KL    6        0.6401281       0.9618317 295.2044 0.4246425
   silhouette.coef silhouette.basis residuals niter   cpu cpu.all nrun cophenetic dispersion
1:       1.0000000        1.0000000  765.7326   450 0.379   6.383   10  0.9936913  0.9819770
2:       0.9751820        0.8990437  657.4086  1440 1.301   7.391   10  0.9971573  0.9512791
3:       0.9414904        0.8710927  583.8487  1740 1.266   8.889   10  0.9974703  0.9614643
4:       0.9378933        0.8119763  531.2162  1540 1.733   8.211   10  0.9978038  0.9677298
5:       0.9458228        0.7543730  481.7107  1490 1.366  10.175   10  0.9980378  0.8799023
   silhouette.consensus
1:            0.9881866
2:            0.8455868
3:            0.8162978
4:            0.8070064
5:            0.7949587
Using NA as a best-fit rank based on decreasing cophenetic correlation coefficient.
Error in if (rank < 1) fstop("invalid argument 'rank': must be greater than 0") : 
  missing value where TRUE/FALSE needed

**Comment by lxilxi** [comment]
Also I have another question:  When estimating the best rank, why the rank always starts from 2, not 1?  Thanks!

**Comment by PoisonAlien** [comment]
Hi,

1. Regarding missing signatures, I am not sure which signatures are missing - what I said was purely a guess on what might be happening. 
2. For the error part, I see that many conversion events have zero mutations. It looks like sample size is way too low or mutation rate per sample is much lesser. In that case the function tries to small positive value to mitigate the issue but it appears that its not helping here. Try adding more samples (I guess you're running this on a subset of samples - try on whole cohort).
3. why not 1 ? because 1 signature is basically all 96 substitution. We want to see if there are any patterns within these 96 substitution, and if that is the case we will start from minimum 2 to guess the best number of patterns.

Hope this helps.

**Comment by lxilxi** [comment]
Thank you very much for the response.  
The set I was using is the whole cohort, we don't have extra more samples/mutations.  The same set I got reasonable results using the version before the reproducible fix, but now it's not. 
I have another question: why sometime it gives the same signature more than once in one output but with different cosine-similarity number? eg. 
> Found Signature_1 most similar to validated Signature_X. Aetiology: ... [cosine-similarity: 0.583] 
Found Signature_4 most similar to validated Signature_X. Aetiology: ... [cosine-similarity: 0.681] 

Thank you again for your help!
